src/TableFunctions/TableFunctionMerge.cpp

```diff
@@ -49,58 +49,74 @@ void TableFunctionMerge::parseArguments(const ASTPtr & ast_function, ContextPtr
             " - name of source database and regexp for table names.",
             ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
 
-    args[0] = evaluateConstantExpressionForDatabaseName(args[0], context);
-    args[1] = evaluateConstantExpressionAsLiteral(args[1], context);
+    auto db_arg = evaluateConstantExpressionForDatabaseName(args[0], context);
+    auto table_arg = evaluateConstantExpressionAsLiteral(args[1], context);
 
-    source_database = args[0]->as<ASTLiteral &>().value.safeGet<String>();
-    source_table_regexp = args[1]->as<ASTLiteral &>().value.safeGet<String>();
+    source_database_regexp = db_arg->as<ASTLiteral &>().value.safeGet<String>();
+    source_table_regexp = table_arg->as<ASTLiteral &>().value.safeGet<String>();
+
+    /// If database argument is not String literal, we should not treat it as regexp
+    if (!args[0]->as<ASTLiteral>())
+        source_database_regexp = "^" + source_database_regexp + "$";
 }
 
 
-const Strings & TableFunctionMerge::getSourceTables(ContextPtr context) const
+const std::unordered_map<String, std::unordered_set<String>> & TableFunctionMerge::getSourceDatabasesAndTables(ContextPtr context) const
 {
-    if (source_tables)
-        return *source_tables;
+    if (source_databases_and_tables)
+        return *source_databases_and_tables;
+
 
-    auto database = DatabaseCatalog::instance().getDatabase(source_database);
+    OptimizedRegularExpression database_re(source_database_regexp);
+    OptimizedRegularExpression table_re(source_table_regexp);
 
-    OptimizedRegularExpression re(source_table_regexp);
-    auto table_name_match = [&](const String & table_name_) { return re.match(table_name_); };
+    auto table_name_match = [&](const String & table_name_) { return table_re.match(table_name_); };
 
     auto access = context->getAccess();
-    bool granted_show_on_all_tables = access->isGranted(AccessType::SHOW_TABLES, source_database);
-    bool granted_select_on_all_tables = access->isGranted(AccessType::SELECT, source_database);
 
-    source_tables.emplace();
-    for (auto it = database->getTablesIterator(context, table_name_match); it->isValid(); it->next())
+    auto databases = DatabaseCatalog::instance().getDatabases();
+
+    for (const auto & db : databases)
     {
-        if (!it->table())
-            continue;
-        bool granted_show = granted_show_on_all_tables || access->isGranted(AccessType::SHOW_TABLES, source_database, it->name());
-        if (!granted_show)
-            continue;
-        if (!granted_select_on_all_tables)
-            access->checkAccess(AccessType::SELECT, source_database, it->name());
-        source_tables->emplace_back(it->name());
+        if (database_re.match(db.first))
```

Comment:
`re.match()` in this case does a aprtial match, IMO when searching for a DB or table, full match is more appropriate. So you might want to modify `OptimizedRegularExpressionImpl<thread_safe>::match` to allow full match, maybe as a last  optional  argument? something like
```
enum class MatchType {FULL, PARTIAL};

bool match(const char * subject, size_t subject_size, Match & match, MatchType match_type = MatchType::PARTIAL) const;
```